File: api/routes/trips.py
"""Маршруты поездок."""

from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, Response

from api.auth.service import AuthError, require_user_llm_config
from api.deps import get_current_user, get_run_manager, get_trip_service
from api.schemas.requests import (
    CreateTripRequest,
    GeocodeRequest,
    ItemFeedbackRequest,
    ReverseGeocodeRequest,
    StartRunRequest,
    UpdatePreferencesRequest,
)
from api.schemas.responses import (
    CityCenterResponse,
    CreateTripResponse,
    GeocodeResponse,
    GeocodeResultResponse,
    ProgramItemResponse,
    ProgramResponse,
    ProgramSectionResponse,
    ReverseGeocodeResponse,
    StructuredProgramResponse,
    TripDetailResponse,
    TripSummaryResponse,
)
from db.users import User
from onboarding.preferences import TripPreferences, merge_trip_preferences, normalize_trip_preferences
from services.run_manager import RunManager
from services.trip_service import ProgramView, TripService

router = APIRouter(prefix="/trips", tags=["trips"])


def _geocode_response(query: str, city_hint: str) -> GeocodeResponse:
    from search.yandex.client import geocode_places as yandex_geocode

    features = yandex_geocode(query, city_hint=city_hint, results=5)
    results: list[GeocodeResultResponse] = []
    for feature in features:
        coords = feature.get("geometry", {}).get("coordinates") or []
        if len(coords) < 2:
            continue
        lon, lat = float(coords[0]), float(coords[1])
        props = feature.get("properties") or {}
        meta = props.get("CompanyMetaData") or {}
        label = str(meta.get("address") or props.get("name") or query).strip()
        results.append(GeocodeResultResponse(lat=lat, lon=lon, label=label))
    if not results and city_hint:
        from search.osm.nominatim import resolve_city_center

        center = resolve_city_center(f"{query}, {city_hint}")
        if center is not None:
            results.append(
                GeocodeResultResponse(
                    lat=center.lat,
                    lon=center.lon,
                    label=center.display_name or query,
                )
            )
    return GeocodeResponse(results=results)


def _reverse_geocode_response(lat: float, lon: float, city_hint: str = "") -> ReverseGeocodeResponse:
    from search.osm.nominatim import reverse_geocode_label as nominatim_reverse
    from search.yandex.client import reverse_geocode_label as yandex_reverse

    label = yandex_reverse(lat, lon, city_hint=city_hint)
    if not label:
        label = nominatim_reverse(lat, lon)
    if not label:
        label = f"{lat:.5f}, {lon:.5f}"
    return ReverseGeocodeResponse(lat=lat, lon=lon, label=label)


def _llm_key_http_error(exc: AuthError) -> HTTPException:
    return HTTPException(
        status_code=exc.status_code,
        detail={"code": "llm_key_required", "message": str(exc)},
    )


def _map_section(view: ProgramView, key: str) -> ProgramSectionResponse:
    section = view.sections.get(key)  # type: ignore[arg-type]
    if section is None:
        return ProgramSectionResponse(intro="", items=[])
    return ProgramSectionResponse(
        intro=section.intro,
        items=[
            ProgramItemResponse(
                index=i.index,
                item_key=i.item_key,
                text=i.text,
                vote=i.vote,
                poi_id=i.poi_id,
            )
            for i in section.items
        ],
    )


def _program_response(view: ProgramView) -> ProgramResponse:
    program = view.program
    sections = StructuredProgramResponse(
        routes=_map_section(view, "routes"),
        route_stops=_map_section(view, "route_stops"),
        lifehacks=_map_section(view, "lifehacks"),
    )
    return ProgramResponse(
        version=view.version,
        version_id=view.version_id,
        scope=view.scope,
        approved=view.approved,
        program=program,
        sections=sections,
        data_warnings=list(view.data_warnings),
        city_fact_status=view.city_fact_status,
    )


@router.get("", response_model=list[TripSummaryResponse])
def list_trips(
    user: User = Depends(get_current_user),
    service: TripService = Depends(get_trip_service),
) -> list[TripSummaryResponse]:
    return [
        TripSummaryResponse(
            id=t.id,
            city=t.city,
            updated_at=t.updated_at,
        )
        for t in service.list_all_trips(user.id)
    ]


@router.post("/geocode", response_model=GeocodeResponse)
def geocode_query(
    body: GeocodeRequest,
    user: User = Depends(get_current_user),
) -> GeocodeResponse:
    """Геокодинг адреса (мастер новой поездки, city_hint обязателен)."""
    _ = user
    return _geocode_response(body.query.strip(), body.city_hint.strip())


@router.post("/reverse-geocode", response_model=ReverseGeocodeResponse)
def reverse_geocode_query(
    body: ReverseGeocodeRequest,
    user: User = Depends(get_current_user),
) -> ReverseGeocodeResponse:
    """Обратный геокодинг для мастера новой поездки."""
    _ = user
    return _reverse_geocode_response(body.lat, body.lon, body.city_hint.strip())


@router.post("", response_model=CreateTripResponse, status_code=201)
def create_trip(
    body: CreateTripRequest,
    user: User = Depends(get_current_user),
    service: TripService = Depends(get_trip_service),
    run_manager: RunManager = Depends(get_run_manager),
) -> CreateTripResponse:
    raw_preferences = (body.preferences.model_dump() if body.preferences else {}) | {
        "route_anchor": body.route_anchor.model_dump() if body.route_anchor else None
    }
    preferences = normalize_trip_preferences(raw_preferences)
    llm_config = None
    if body.start_run:
        try:
            llm_config = require_user_llm_config(user.id)
        except AuthError as exc:
            raise _llm_key_http_error(exc) from exc
    trip_id = service.create_new_trip(
        city=body.city,
        dates="Без дат",
        origin_city=body.city,
        user_query=body.user_query,
        preferences=preferences,
        user_id=user.id,
    )
    run_id: str | None = None
    if body.start_run:
        assert llm_config is not None
        details = service.get_trip_details(trip_id, user_id=user.id)
        assert details is not None
        trip = details.trip
        state = service.build_initial_state(
            trip_id=trip_id,
            city=trip["city"],
            dates=trip["dates"],
            origin_city=trip["origin_city"],
            search_context=service.apply_preferences(preferences),
            preferences_dict=preferences.model_dump(),
            rebuild_scope="full",
            user_message=trip.get("user_query") or body.user_query,
        )
        run_id = run_manager.start_run(state, llm_config=llm_config)
    return CreateTripResponse(trip_id=trip_id, run_id=run_id)


@router.delete("/{trip_id}", status_code=204, response_class=Response)
def delete_trip(
    trip_id: int,
    user: User = Depends(get_current_user),
    service: TripService = Depends(get_trip_service),
    run_manager: RunManager = Depends(get_run_manager),
) -> Response:
    try:
        service.delete_trip_by_id(
            trip_id,
            user_id=user.id,
            has_active_run=run_manager.has_active_run_for_trip(trip_id),
        )
    except ValueError as exc:
        message = str(exc)
        if "сборки" in message:
            raise HTTPException(status_code=409, detail=message) from exc
        raise HTTPException(status_code=404, detail=message) from exc
    run_manager.forget_runs_for_trip(trip_id)
    return Response(status_code=204)


@router.get("/{trip_id}", response_model=TripDetailResponse)
def get_trip(
    trip_id: int,
    user: User = Depends(get_current_user),
    service: TripService = Depends(get_trip_service),
) -> TripDetailResponse:
    details = service.get_trip_details(trip_id, user_id=user.id)
    if details is None:
        raise HTTPException(status_code=404, detail="Поездка не найдена")
    trip = details.trip
    return TripDetailResponse(
        id=int(trip["id"]),
        city=trip["city"],
        user_query=trip.get("user_query"),
        created_at=trip["created_at"],
        updated_at=trip["updated_at"],
    )


@router.get("/{trip_id}/program", response_model=ProgramResponse)
def get_program(
    trip_id: int,
    user: User = Depends(get_current_user),
    service: TripService = Depends(get_trip_service),
) -> ProgramResponse:
    if service.get_trip_details(trip_id, user_id=user.id) is None:
        raise HTTPException(status_code=404, detail="Поездка не найдена")
    view = service.get_program_view(trip_id, user_id=user.id)
    if view is None:
        raise HTTPException(status_code=404, detail="Программа не найдена")
    return _program_response(view)


@router.put("/{trip_id}/preferences", response_model=TripPreferences)
def update_preferences(
    trip_id: int,
    body: UpdatePreferencesRequest,
    user: User = Depends(get_current_user),
    service: TripService = Depends(get_trip_service),
) -> TripPreferences:
    if service.get_trip_details(trip_id, user_id=user.id) is None:
        raise HTTPException(status_code=404, detail="Поездка не найдена")
    update = body.model_dump(exclude_unset=True)
    if not update:
        raise HTTPException(status_code=400, detail="Нет полей для обновления")
    try:
        return service.update_trip_preferences(trip_id, user_id=user.id, update=update)
    except ValueError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc


@router.get("/{trip_id}/city-center", response_model=CityCenterResponse)
def get_city_center(
    trip_id: int,
    user: User = Depends(get_current_user),
    service: TripService = Depends(get_trip_service),
) -> CityCenterResponse:
    details = service.get_trip_details(trip_id, user_id=user.id)
    if details is None:
        raise HTTPException(status_code=404, detail="Поездка не найдена")
    from search.osm.nominatim import resolve_city_center

    city = str(details.trip["city"])
    center = resolve_city_center(city)
    if center is None:
        raise HTTPException(status_code=404, detail=f"Не удалось определить центр города: {city}")
    return CityCenterResponse(
        lat=center.lat,
        lon=center.lon,
        label=center.display_name or city,
    )


@router.post("/{trip_id}/geocode", response_model=GeocodeResponse)
def geocode_address(
    trip_id: int,
    body: GeocodeRequest,
    user: User = Depends(get_current_user),
    service: TripService = Depends(get_trip_service),
) -> GeocodeResponse:
    details = service.get_trip_details(trip_id, user_id=user.id)
    if details is None:
        raise HTTPException(status_code=404, detail="Поездка не найдена")
    city_hint = (body.city_hint or str(details.trip["city"])).strip()
    return _geocode_response(body.query.strip(), city_hint)


@router.post("/{trip_id}/reverse-geocode", response_model=ReverseGeocodeResponse)
def reverse_geocode_address(
    trip_id: int,
    body: ReverseGeocodeRequest,
    user: User = Depends(get_current_user),
    service: TripService = Depends(get_trip_service),
) -> ReverseGeocodeResponse:
    details = service.get_trip_details(trip_id, user_id=user.id)
    if details is None:
        raise HTTPException(status_code=404, detail="Поездка не найдена")
    city_hint = (body.city_hint or str(details.trip["city"])).strip()
    return _reverse_geocode_response(body.lat, body.lon, city_hint)


@router.put("/{trip_id}/program/feedback", response_model=ProgramResponse)
def set_program_feedback(
    trip_id: int,
    body: ItemFeedbackRequest,
    user: User = Depends(get_current_user),
    service: TripService = Depends(get_trip_service),
) -> ProgramResponse:
    if service.get_trip_details(trip_id, user_id=user.id) is None:
        raise HTTPException(status_code=404, detail="Поездка не найдена")
    try:
        service.set_item_feedback(
            trip_id,
            version_id=body.version_id,
            section=body.section,
            item_key=body.item_key,
            item_index=body.item_index,
            vote=body.vote,
        )
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    view = service.get_program_view(trip_id, user_id=user.id)
    if view is None:
        raise HTTPException(status_code=404, detail="Программа не найдена")
    return _program_response(view)


@router.get("/{trip_id}/preferences", response_model=TripPreferences | None)
def get_preferences(
    trip_id: int,
    user: User = Depends(get_current_user),
    service: TripService = Depends(get_trip_service),
) -> TripPreferences | None:
    details = service.get_trip_details(trip_id, user_id=user.id)
    if details is None:
        raise HTTPException(status_code=404, detail="Поездка не найдена")
    if details.preferences is None:
        return None
    return TripPreferences.model_validate(details.preferences)


@router.post("/{trip_id}/runs", response_model=CreateTripResponse)
def start_run(
    trip_id: int,
    body: StartRunRequest,
    user: User = Depends(get_current_user),
    service: TripService = Depends(get_trip_service),
    run_manager: RunManager = Depends(get_run_manager),
) -> CreateTripResponse:
    if service.get_trip_details(trip_id, user_id=user.id) is None:
        raise HTTPException(status_code=404, detail="Поездка не найдена")
    try:
        llm_config = require_user_llm_config(user.id)
        state = service.prepare_continue_trip(trip_id, body.scope)
    except AuthError as exc:
        raise _llm_key_http_error(exc) from exc
    except ValueError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    run_id = run_manager.start_run(state, llm_config=llm_config)
    return CreateTripResponse(trip_id=trip_id, run_id=run_id)

File: api-node/src/routes/trips.ts
import { randomUUID } from "node:crypto";
import type { FastifyInstance, FastifyReply } from "fastify";
import { z } from "zod";
import { requireAuth } from "../middleware/auth.js";
import {
  mergeTripPreferences,
  normalizeTripPreferences,
  routeAnchorSchema,
} from "../lib/preferences.js";
import {
  AuthError,
  assertCanStartRun,
} from "../services/auth.js";
import { feedbackSchema, setItemFeedback } from "../services/itemFeedback.js";
import { FreeRunQuotaError } from "../services/freeQuotas.js";
import { RunQuotaError } from "../services/quotas.js";
import {
  hasActiveRunForTrip,
  startRun,
} from "../services/runManager.js";
import { buildProgramView } from "../services/programView.js";
import { buildTripOsrmUpdateStatus } from "../services/osrmTripUpdate.js";
import { recoverCityFactIfNeeded } from "../services/cityFactRecovery.js";
import { repairProgramForTrip } from "../services/repairProgram.js";
import {
  geocodePlaces,
  resolveCityCenter,
  reverseGeocodeLabel,
} from "../services/geocode.js";
import { recordAuditEvent } from "../repos/audit.js";
import { saveUserProfile } from "../repos/users.js";
import {
  bearerSecurity,
  createTripBodySchema,
  geocodeBodySchema,
  ref,
  startRunBodySchema,
} from "../openapi/components.js";

import * as tripsRepo from "../repos/trips.js";
import {
  InputValidationError,
  sanitizeAndValidate,
} from "../lib/inputValidation.js";
import { normalizePoiFactCacheKey } from "../lib/poiFactCacheKey.js";
import { enqueuePoiFact } from "../jobs/enqueue.js";
import * as poiFactsRepo from "../repos/poiFacts.js";

const startRunSchema = startRunBodySchema;
const createTripSchema = createTripBodySchema;

const updatePrefsSchema = z
  .object({
    travel_party: z
      .enum([
        "solo",
        "couple",
        "family",
        "friends",
        "parent_child",
        "family_two",
      ])
      .optional(),
    route_anchor: routeAnchorSchema.nullable().optional(),
  })
  .refine((v) => Object.keys(v).length > 0, {
    message: "Нет полей для обновления",
  });

const geocodeSchema = geocodeBodySchema;

const reverseGeocodeSchema = z.object({
  lat: z.number().min(-90).max(90),
  lon: z.number().min(-180).max(180),
  city_hint: z.string().max(128).default(""),
});

const poiFactStartSchema = z.object({
  poi_id: z.string().max(128).optional().nullable(),
  name: z.string().min(1).max(256),
});

function validationErrorReply(reply: FastifyReply, err: InputValidationError) {
  return reply.code(400).send({ detail: err.message });
}

export async function registerTripsRoutes(app: FastifyInstance): Promise<void> {
  app.get(
    "/api/trips",
    {
      preHandler: requireAuth,
      schema: {
        tags: ["trips"],
        summary: "List trips",
        security: [...bearerSecurity],
        response: {
          200: { type: "array", items: ref("TripSummary") },
        },
      },
    },
    async (request) => {
      const rows = await tripsRepo.listTrips(request.user!.id);
      return rows;
    },
  );

  app.post(
    "/api/trips/geocode",
    {
      preHandler: requireAuth,
      schema: {
        tags: ["trips"],
        summary: "Geocode (new trip wizard)",
        security: [...bearerSecurity],
        body: ref("GeocodeRequest"),
      },
    },
    async (request, reply) => {
      const body = geocodeSchema.safeParse(request.body);
      if (!body.success) {
        return reply.code(400).send({ detail: "Некорректные данные" });
      }
      const results = await geocodePlaces(
        body.data.query.trim(),
        body.data.city_hint.trim(),
      );
      return { results };
    },
  );

  app.post(
    "/api/trips/reverse-geocode",
    { preHandler: requireAuth },
    async (request, reply) => {
      const body = reverseGeocodeSchema.safeParse(request.body);
      if (!body.success) {
        return reply.code(400).send({ detail: "Некорректные данные" });
      }
      let label = await reverseGeocodeLabel(
        body.data.lat,
        body.data.lon,
        body.data.city_hint.trim(),
      );
      if (!label) {
        label = `${body.data.lat.toFixed(5)}, ${body.data.lon.toFixed(5)}`;
      }
      return { lat: body.data.lat, lon: body.data.lon, label };
    },
  );

  app.post(
    "/api/trips",
    {
      preHandler: requireAuth,
      schema: {
        tags: ["trips"],
        summary: "Create trip",
        security: [...bearerSecurity],
        body: ref("CreateTripRequest"),
        response: {
          201: ref("CreateTripResponse"),
          400: ref("ErrorDetail"),
          409: ref("ErrorDetail"),
          428: ref("ErrorDetail"),
          429: ref("ErrorDetail"),
          503: ref("ErrorDetail"),
        },
      },
    },
    async (request, reply) => {
      const body = createTripSchema.safeParse(request.body);
      if (!body.success) {
        return reply.code(400).send({ detail: "Некорректные данные" });
      }
      const rawPrefs = {
        ...(body.data.preferences ?? {}),
        route_anchor: body.data.route_anchor ?? null,
      };
      const preferences = normalizeTripPreferences(rawPrefs);
      if (body.data.start_run) {
        try {
          await assertCanStartRun(request.user!.id);
        } catch (err) {
          if (err instanceof AuthError) {
            const code =
              err.statusCode === 503 ? "ai_platform_unavailable" : "llm_key_required";
            const status = err.statusCode === 503 ? 503 : 428;
            return reply.code(status).send({
              detail: { code, message: err.message },
            });
          }
          throw err;
        }
      }
      let city: string;
      let userQuery: string;
      try {
        city = sanitizeAndValidate(body.data.city, "city");
        userQuery = sanitizeAndValidate(body.data.user_query, "message");
      } catch (err) {
        if (err instanceof InputValidationError) {
          return validationErrorReply(reply, err);
        }
        throw err;
      }
      const tripId = await tripsRepo.createTrip({
        userId: request.user!.id,
        city,
        dates: "Без дат",
        originCity: city,
        userQuery,
      });
      await tripsRepo.savePreferences(tripId, preferences);
      await saveUserProfile(request.user!.id, preferences);
      await recordAuditEvent({
        action: "trip.create",
        entityType: "trip",
        entityId: String(tripId),
        userId: request.user!.id,
        metadata: { city },
      });

      let runId: string | null = null;
      if (body.data.start_run) {
        try {
          runId = await startRun(tripId, "full");
        } catch (err) {
          if (err instanceof FreeRunQuotaError) {
            return reply.code(429).send({
              detail: { code: "free_run_quota_exceeded", message: err.message },
            });
          }
          if (err instanceof RunQuotaError) {
            return reply.code(429).send({
              detail: { code: "run_quota_exceeded", message: err.message },
            });
          }
          if (err instanceof Error && /сборк|маршрут/i.test(err.message)) {
            return reply.code(409).send({
              detail: { code: "active_run", message: err.message },
            });
          }
          throw err;
        }
      }
      return reply.code(201).send({ trip_id: tripId, run_id: runId });
    },
  );

  app.delete<{ Params: { trip_id: string } }>(
    "/api/trips/:trip_id",
    { preHandler: requireAuth },
    async (request, reply) => {
      const tripId = Number(request.params.trip_id);
      const trip = await tripsRepo.getTrip(tripId, request.user!.id);
      if (!trip) {
        return reply.code(404).send({ detail: "Поездка не найдена" });
      }
      if (await hasActiveRunForTrip(tripId)) {
        return reply
          .code(409)
          .send({ detail: "Нельзя удалить поездку во время сборки программы" });
      }
      const ok = await tripsRepo.deleteTrip(tripId, request.user!.id);
      if (!ok) {
        return reply.code(404).send({ detail: "Поездка не найдена" });
      }
      return reply.code(204).send();
    },
  );

  app.get<{ Params: { trip_id: string } }>(
    "/api/trips/:trip_id",
    { preHandler: requireAuth },
    async (request, reply) => {
      const trip = await tripsRepo.getTrip(
        Number(request.params.trip_id),
        request.user!.id,
      );
      if (!trip) {
        return reply.code(404).send({ detail: "Поездка не найдена" });
      }
      return {
        id: trip.id,
        city: trip.city,
        user_query: trip.user_query,
        created_at: trip.created_at,
        updated_at: trip.updated_at,
      };
    },
  );

  app.get<{ Params: { trip_id: string } }>(
    "/api/trips/:trip_id/program",
    { preHandler: requireAuth },
    async (request, reply) => {
      const tripId = Number(request.params.trip_id);
      const trip = await tripsRepo.getTrip(tripId, request.user!.id);
      if (!trip) {
        return reply.code(404).send({ detail: "Поездка не найдена" });
      }
      const latest = await tripsRepo.getLatestItinerary(tripId);
      if (!latest) {
        return reply.code(404).send({ detail: "Программа не найдена" });
      }
      const repaired = await repairProgramForTrip(tripId, trip, latest.program);
      const program = await recoverCityFactIfNeeded({
        tripId,
        userId: request.user!.id,
        city: trip.city,
        versionId: latest.id,
        program: repaired,
      });
      return buildProgramView(tripId, { ...latest, program });
    },
  );

  app.get<{ Params: { trip_id: string } }>(
    "/api/trips/:trip_id/osrm-update",
    {
      preHandler: requireAuth,
      schema: {
        tags: ["trips"],
        summary: "Есть ли более новый OSRM-граф, чем последняя сборка маршрутов",
      },
    },
    async (request, reply) => {
      const tripId = Number(request.params.trip_id);
      const trip = await tripsRepo.getTrip(tripId, request.user!.id);
      if (!trip) {
        return reply.code(404).send({ detail: "Поездка не найдена" });
      }
      const latest = await tripsRepo.getLatestItinerary(tripId);
      return buildTripOsrmUpdateStatus({ city: trip.city, latest });
    },
  );

  app.put<{ Params: { trip_id: string } }>(
    "/api/trips/:trip_id/preferences",
    { preHandler: requireAuth },
    async (request, reply) => {
      const tripId = Number(request.params.trip_id);
      const trip = await tripsRepo.getTrip(tripId, request.user!.id);
      if (!trip) {
        return reply.code(404).send({ detail: "Поездка не найдена" });
      }
      const body = updatePrefsSchema.safeParse(request.body);
      if (!body.success) {
        return reply
          .code(400)
          .send({ detail: body.error.errors[0]?.message ?? "Некорректные данные" });
      }
      const existing = await tripsRepo.getPreferences(tripId);
      const merged = mergeTripPreferences(existing, body.data);
      await tripsRepo.savePreferences(tripId, merged);
      return merged;
    },
  );

  app.get<{ Params: { trip_id: string } }>(
    "/api/trips/:trip_id/preferences",
    { preHandler: requireAuth },
    async (request, reply) => {
      const tripId = Number(request.params.trip_id);
      const trip = await tripsRepo.getTrip(tripId, request.user!.id);
      if (!trip) {
        return reply.code(404).send({ detail: "Поездка не найдена" });
      }
      const prefs = await tripsRepo.getPreferences(tripId);
      return prefs ? normalizeTripPreferences(prefs) : null;
    },
  );

  app.get<{ Params: { trip_id: string } }>(
    "/api/trips/:trip_id/city-center",
    { preHandler: requireAuth },
    async (request, reply) => {
      const trip = await tripsRepo.getTrip(
        Number(request.params.trip_id),
        request.user!.id,
      );
      if (!trip) {
        return reply.code(404).send({ detail: "Поездка не найдена" });
      }
      const center = await resolveCityCenter(trip.city);
      if (!center) {
        return reply
          .code(404)
          .send({ detail: `Не удалось определить центр города: ${trip.city}` });
      }
      return center;
    },
  );

  app.post<{ Params: { trip_id: string } }>(
    "/api/trips/:trip_id/geocode",
    { preHandler: requireAuth },
    async (request, reply) => {
      const tripId = Number(request.params.trip_id);
      const trip = await tripsRepo.getTrip(tripId, request.user!.id);
      if (!trip) {
        return reply.code(404).send({ detail: "Поездка не найдена" });
      }
      const body = geocodeSchema.safeParse(request.body);
      if (!body.success) {
        return reply.code(400).send({ detail: "Некорректные данные" });
      }
      const cityHint = (body.data.city_hint || trip.city).trim();
      const results = await geocodePlaces(body.data.query.trim(), cityHint);
      return { results };
    },
  );

  app.post<{ Params: { trip_id: string } }>(
    "/api/trips/:trip_id/reverse-geocode",
    { preHandler: requireAuth },
    async (request, reply) => {
      const tripId = Number(request.params.trip_id);
      const trip = await tripsRepo.getTrip(tripId, request.user!.id);
      if (!trip) {
        return reply.code(404).send({ detail: "Поездка не найдена" });
      }
      const body = reverseGeocodeSchema.safeParse(request.body);
      if (!body.success) {
        return reply.code(400).send({ detail: "Некорректные данные" });
      }
      const cityHint = (body.data.city_hint || trip.city).trim();
      let label = await reverseGeocodeLabel(
        body.data.lat,
        body.data.lon,
        cityHint,
      );
      if (!label) {
        label = `${body.data.lat.toFixed(5)}, ${body.data.lon.toFixed(5)}`;
      }
      return { lat: body.data.lat, lon: body.data.lon, label };
    },
  );

  app.post<{ Params: { trip_id: string } }>(
    "/api/trips/:trip_id/runs",
    {
      preHandler: requireAuth,
      schema: {
        tags: ["trips"],
        summary: "Start graph run",
        security: [...bearerSecurity],
        body: ref("StartRunRequest"),
      },
    },
    async (request, reply) => {
      const tripId = Number(request.params.trip_id);
      const trip = await tripsRepo.getTrip(tripId, request.user!.id);
      if (!trip) {
        return reply.code(404).send({ detail: "Поездка не найдена" });
      }
      const body = startRunSchema.safeParse(request.body ?? {});
      if (!body.success) {
        return reply.code(400).send({ detail: "Некорректные данные" });
      }
      try {
        await assertCanStartRun(request.user!.id);
        const runId = await startRun(tripId, body.data.scope);
        return { trip_id: tripId, run_id: runId };
      } catch (err) {
        if (err instanceof AuthError) {
          const code =
            err.statusCode === 503 ? "ai_platform_unavailable" : "llm_key_required";
          const status = err.statusCode === 503 ? 503 : 428;
          return reply.code(status).send({
            detail: { code, message: err.message },
          });
        }
        if (err instanceof FreeRunQuotaError) {
          return reply.code(429).send({
            detail: { code: "free_run_quota_exceeded", message: err.message },
          });
        }
        if (err instanceof RunQuotaError) {
          return reply.code(429).send({
            detail: { code: "run_quota_exceeded", message: err.message },
          });
        }
        if (err instanceof Error && /сборк|маршрут/i.test(err.message)) {
          return reply.code(409).send({
            detail: { code: "active_run", message: err.message },
          });
        }
        throw err;
      }
    },
  );

  app.put<{ Params: { trip_id: string } }>(
    "/api/trips/:trip_id/program/feedback",
    { preHandler: requireAuth },
    async (request, reply) => {
      const tripId = Number(request.params.trip_id);
      const trip = await tripsRepo.getTrip(tripId, request.user!.id);
      if (!trip) {
        return reply.code(404).send({ detail: "Поездка не найдена" });
      }
      const body = feedbackSchema.safeParse(request.body);
      if (!body.success) {
        return reply.code(400).send({ detail: body.error.errors[0]?.message });
      }
      try {
        await setItemFeedback(tripId, body.data);
      } catch (err) {
        return reply
          .code(400)
          .send({ detail: err instanceof Error ? err.message : String(err) });
      }
      const latest = await tripsRepo.getLatestItinerary(tripId);
      if (!latest) {
        return reply.code(404).send({ detail: "Программа не найдена" });
      }
      const repaired = await repairProgramForTrip(tripId, trip, latest.program);
      return buildProgramView(tripId, { ...latest, program: repaired });
    },
  );

  app.post<{ Params: { trip_id: string } }>(
    "/api/trips/:trip_id/poi-facts",
    { preHandler: requireAuth },
    async (request, reply) => {
      const tripId = Number(request.params.trip_id);
      const trip = await tripsRepo.getTrip(tripId, request.user!.id);
      if (!trip) {
        return reply.code(404).send({ detail: "Поездка не найдена" });
      }
      const body = poiFactStartSchema.safeParse(request.body ?? {});
      if (!body.success) {
        return reply.code(400).send({ detail: "Некорректные данные" });
      }
      const poiId = (body.data.poi_id ?? "").trim();
      const name = body.data.name.trim();
      const cacheKey = normalizePoiFactCacheKey({
        poiId: poiId || null,
        name,
        city: trip.city,
      });

      const existing = await poiFactsRepo.getPoiFact(cacheKey);
      if (
        existing?.status === "ready" &&
        existing.text &&
        !poiFactsRepo.looksLikeSearchGarbage(existing.text)
      ) {
        return poiFactsRepo.toPoiFactResponse(existing);
      }
      if (
        existing?.status === "pending" &&
        !poiFactsRepo.isPendingStale(existing.updated_at)
      ) {
        return poiFactsRepo.toPoiFactResponse(existing);
      }

      try {
        await assertCanStartRun(request.user!.id);
      } catch (err) {
        if (err instanceof AuthError) {
          const code =
            err.statusCode === 503 ? "ai_platform_unavailable" : "llm_key_required";
          const status = err.statusCode === 503 ? 503 : 428;
          return reply.code(status).send({
            detail: { code, message: err.message },
          });
        }
        throw err;
      }

      const sourceKind = poiId
        ? poiId.startsWith("Q") || poiId.startsWith("wikidata_")
          ? "wikidata"
          : /^osm_/.test(poiId)
            ? "osm"
            : "search"
        : "search";

      const row = await poiFactsRepo.upsertPoiFactPending({
        cacheKey,
        poiName: name,
        city: trip.city,
        sourceKind,
      });

      await enqueuePoiFact(randomUUID(), {
        trip_id: tripId,
        user_id: request.user!.id,
        city: trip.city,
        cache_key: cacheKey,
        poi_id: poiId,
        name,
      });

      return poiFactsRepo.toPoiFactResponse(row);
    },
  );

  app.get<{ Params: { trip_id: string; cache_key: string } }>(
    "/api/trips/:trip_id/poi-facts/:cache_key",
    { preHandler: requireAuth },
    async (request, reply) => {
      const tripId = Number(request.params.trip_id);
      const trip = await tripsRepo.getTrip(tripId, request.user!.id);
      if (!trip) {
        return reply.code(404).send({ detail: "Поездка не найдена" });
      }
      const cacheKey = decodeURIComponent(request.params.cache_key).trim();
      if (!cacheKey) {
        return reply.code(400).send({ detail: "Некорректный cache_key" });
      }
      const row = await poiFactsRepo.getPoiFact(cacheKey);
      if (!row) {
        return reply.code(404).send({ detail: "Справка не найдена" });
      }
      if (
        row.status === "ready" &&
        row.text &&
        poiFactsRepo.looksLikeSearchGarbage(row.text)
      ) {
        return {
          cache_key: row.cache_key,
          name: row.poi_name,
          status: "failed" as const,
          text: null,
          error: poiFactsRepo.POI_FACT_NOT_FOUND,
        };
      }
      return poiFactsRepo.toPoiFactResponse(row);
    },
  );
}
